pkg/relay/register.go

```diff
@@ -0,0 +1,296 @@
+	Pubkey     types.PublicKey
+	ID         int
+	Response   chan Resp
+}
+
+// Resp respone structure
+// - potential candidate for structure pool
+// as it's almost constant size
+type Resp struct {
+	ID     int
+	Type   int8
+	Commit bool
+	Err    error
+}
+
+// ***** Builder Domain *****
+// RegisterValidator is called is called by validators communicating through mev-boost who would like to receive a block from us when their slot is scheduled
+func (rs *Relay) RegisterValidator(ctx context.Context, payload []structs.SignedValidatorRegistration) error {
+	timer := prometheus.NewTimer(rs.m.Timing.WithLabelValues("registerValidator", "all"))
+	defer timer.ObserveDuration()
+
+	var respCh chan Resp
+
+	// This function is limitted to the size of the buffer to prevent deadlocks
+	if rs.config.RegisterValidatorMaxNum > uint64(len(payload)) {
+		respCh = rs.retChannPool.Get().(chan Resp)
+		defer rs.retChannPool.Put(respCh)
+	}
+
+	fc := NewFlowControl(respCh, len(payload))
+	defer fc.Close()
+
+	// verify other parameters in separate goroutine
+	go verifyOthers(rs.beaconState, rs.regMngr, fc, payload)
+
+	// synchronize responses from:
+	//	- verification of signatures
+	//	- other verifications of payload validity
+	//	- successfull store
+	go synchAndStore(rs.regMngr, fc, payload)
+
+	verifyChan := rs.regMngr.GetVerifyChan(ResponseQueueRegister)
+SendPayloads:
+	for i, p := range payload {
+		msg, err := types.ComputeSigningRoot(payload[i].Message, rs.config.BuilderSigningDomain)
+		if err != nil {
+			fc.SentVerificationInc()
+			fc.RespCh <- Resp{ID: i, Err: errors.New("invalid signature"), Type: ResponseTypeVerify}
+			break SendPayloads
+		}
+
+		select {
+		case <-fc.FailureCh:
+			break SendPayloads
+		case verifyChan <- VerifyReq{
+			Signature: p.Signature,
+			Pubkey:    p.Message.Pubkey,
+			Msg:       msg,
+			ID:        i,
+			Response:  fc.RespCh}:
+			fc.SentVerificationInc()
+		}
+	}
+	return <-fc.ExitCh
+}
+
+func (rs *Relay) RegisterValidatorSingular(ctx context.Context, payload structs.SignedValidatorRegistration) error {
+	timer := prometheus.NewTimer(rs.m.Timing.WithLabelValues("registerValidatorSingular", "all"))
+	defer timer.ObserveDuration()
+
+	otherChecks, _ := verifyOther(rs.beaconState.Beacon(), rs.regMngr, 0, payload)
+	if otherChecks.Err != nil {
+		return otherChecks.Err
+	}
+	if !otherChecks.Commit { // when timestamp is wrong so we don't commit
+		return nil
+	}
+
+	timer2 := prometheus.NewTimer(rs.m.Timing.WithLabelValues("registerValidatorSingular", "verify"))
+	msg, err := types.ComputeSigningRoot(payload.Message, rs.config.BuilderSigningDomain)
+	if err != nil {
+		return errors.New("invalid signature")
+	}
+
+	respCh := rs.singleRetChannPool.Get().(chan Resp)
+	defer rs.singleRetChannPool.Put(respCh)
+
+	rs.regMngr.GetVerifyChan(ResponseQueueRegister) <- VerifyReq{
+		Signature: payload.Signature,
+		Pubkey:    payload.Message.Pubkey,
+		Msg:       msg,
+		Response:  respCh,
+	}
+	r := <-respCh
```

Comment:
Is this guaranteed not to block?  Can we add a comment explaning why?

Either way, can we select along with `<-ctx.Done()` just to be safe?

Reply:
Makes sense.  SGTM.  Maybe add a small comment saying something like `// guaranteed to receive` ?